src/tregex.ur

```diff
@@ -0,0 +1,342 @@
+    @foldUR
+     [char_class] [fn r => char_class]
+     (fn [nm ::_] [rest ::_] [[nm] ~ rest] (cc : char_class) acc => acc ^ cc)
+     ""
+     fl
+     cs
+
+val c_digit = "\\d"
+val c_whitespace = "\\s"
+val c_word = "\\w"
+val c_boundary = "[\\b]"
+
+fun c_char (c : char) = escape (String.str c)
+
+fun one_of (cc : char_class) : t [] = {Groups = {}, NGroups = 0, Expr = "[" ^ cc ^ "]"}
+fun none_of (cc : char_class) : t [] = {Groups = {}, NGroups = 0, Expr = "[^" ^ cc ^ "]"}
+
+(* ****** ****** *)
+			
+fun projs [keep ::: {Type}] [drop ::: {Type}] [keep ~ drop] (xs : $(keep ++ drop)): $keep = xs --- drop
+
+fun alt [s1 ::: {Unit}] [s2 ::: {Unit}] [s1 ~ s2]
+	(fls1 : folder s1) (fls2 : folder s2)
+	(x: t s1) (y : t s2): t (s1 ++ s2) =
+    {Groups = x.Groups ++ (@shift x.NGroups y.Groups fls2),
+     NGroups = x.NGroups + y.NGroups,
+     Expr = "(?:" ^ x.Expr ^ ")|(?:" ^ y.Expr ^ ")"}
+
+(* ****** ****** *)
+    
+fun capture
+	[r ::: {Unit}] [nm :: Name] [r ~ [nm]]
+	(fl : folder r)
+	(y: t r): t (r ++ [nm]) =
+    {Groups = (@shift 1 y.Groups fl) ++ {nm = 0},
+     NGroups = y.NGroups + 1,
+     Expr = "(" ^ y.Expr ^ ")"}
+
+(* ****** ****** *)
+
+fun groups [r ::: {Unit}] (x : t r) (fl : folder r): $(map (fn _ => int) r) = x.Groups
+
+val show_tsregex = fn [r ::: {Unit}] => mkShow (fn x => x.Expr)
+
+(* ****** ****** *)
+
+type counted_substring = {Start : int, Len : int}					
+con match = fn (r :: {Unit}) (a :: Type) => {Whole : a, Groups : $(map (fn _ => a) r)}
+
+fun substring_offset (delta : int) (substring : counted_substring)
+    : counted_substring =
+    {Start = substring.Start + delta, Len = substring.Len}
+
+fun match_mp [r ::: {Unit}]
+	     [a ::: Type]
+	     [b ::: Type]
+	     (f : a -> b)
+	     (fl : folder r)
+	     (match : match r a) : match r b =
+    {Whole = f match.Whole,
+     Groups =
+     @mp [fn _ => a] [fn _ => b] (fn [t] x => f x) fl match.Groups}
+
+fun match_offset
+	[r ::: {Unit}]
+	(delta : int)
+	(fl : folder r)
+	(m : match r counted_substring)
+    : match r counted_substring =
+    {Whole = substring_offset delta m.Whole,
+     Groups =
+     @mp [fn _ => counted_substring] [fn _ => counted_substring] (fn [t] x => substring_offset delta x) fl m.Groups}
+
+(* Returns the index of the character just _after_ a match. *)
+fun after_match [r ::: {Unit}] (match : match r counted_substring) : int =
+  match.Whole.Start + match.Whole.Len
+
+fun get_substrings [r ::: {Unit}] (fl : folder r)
+		   (haystack : string) (m : match r counted_substring) : match r (option string) =
+    {Whole =
+     if m.Whole.Len = 0 then None
+     else Some (String.substring haystack m.Whole),
+     Groups =
+     @mp
+      [fn _ => counted_substring] [fn _ => option string]
+      (fn [t] x =>
+	  if x.Len = 0 then None
+	  else Some (String.substring haystack x))
+      fl m.Groups}
+
+(* Unmarshaling FFI types *)
+
+structure FFI = Regex__FFI
```

Comment:
How much of this section could you implement on top of `Regex`, rather than calling directly into the FFI module? I’m always in favor of avoiding the FFI if feasible.